File: code/ex255.java
class Solution {
    public boolean verifyPreorder(int[] preorder) {
        if (preorder.length <= 2)    return true;
        Stack<Integer> stack = new Stack<>();
        int root = Integer.MIN_VALUE;
        for (int node : preorder) {
            // Go over the left part of a subtree
            while (!stack.isEmpty() && node > stack.peek()) {
                root = stack.pop();
            }
            if (node < root)    return false;    // Test the right part of a subtree
            stack.push(node);
        }
        return true;
    }
}

File: code/ex255.cpp
class Solution {
public:
    bool verifyPreorder(vector<int>& preorder) {
        int n = preorder.size();
        
        stack<int> stk;
        int root = INT_MIN;
        for (int i = 0; i < n; i++) {
            while (!stk.empty() && preorder[i] > stk.top()) {
                root = stk.top();
                stk.pop();
            }
            if (preorder[i] < root)    return 0;
            stk.push(preorder[i]);
        }
        
        return 1;
    }
};